File: 2019/8/solution.js
const fs = require('fs');
const path = require('path');

const input = fs.readFileSync(path.join(__dirname, 'input.txt'), 'utf-8', (err, data) => {
  if (err) {
    console.log(err)
  } else {
    return data;
  }
});

const width = 25;
const height = 6;
const image = [];

for (let i = 0; i < input.length; i += width * height) {
  const layer = [];
  for (let j = 0; j < height; j++) {
    layer.push(input.slice(i + width * j, i + (width) * (j + 1)));
  }
  image.push(layer);
}

let minZeros = Infinity;
let part1 = 0;

for (const layer of image) {
  let count0 = 0;
  let count1 = 0;
  let count2 = 0;
  for (const row of layer) {
    for (const pixel of row) {
      switch (pixel) {
        case '0':
          count0++;
          break;
        case '1':
          count1++;
          break;
        case '2':
          count2++;
      }
    }
  }
  if (count0 < minZeros) {
    minZeros = count0;
    part1 = count1 * count2;
  }
}

console.log(part1);

const message = [];
for (let row = 0; row < height; row++) {
  const line = [];
  for (let col = 0; col < width; col++) {
    for (const layer of image) {
      if (layer[row][col] === '1') {
        line.push('@');
        break;
      }
      if (layer[row][col] === '0') {
        line.push(' ');
        break;
      }
    }
  }
  message.push(line.join(''));
}

for (const line of message) {
  console.log(line);
}

File: 2019/8/solution.go
package main

import (
	"fmt"
	"math"
	"os"
	"path/filepath"
	"runtime"
	"strings"
)

func main() {
	_, filename, _, _ := runtime.Caller(0)
	dirname := filepath.Dir(filename)
	inputFilePath := filepath.Join(dirname, "input.txt")
	data, err := os.ReadFile(inputFilePath)
	if err != nil {
		panic(err)
	}
	puzzleInput := parseInput(string(data), 25, 6)
	fmt.Println(part1(puzzleInput))
	fmt.Println(part2(puzzleInput))
}

func parseInput(data string, w, h int) [][][]int {
	var image [][][]int
	var layer [][]int
	var row []int
	for i := 0; i*w*h < len(data); i++ {
		layer = make([][]int, h)
		for r := range h {
			row = make([]int, w)
			for c := range w {
				row[c] = int(data[i*w*h+r*w+c] - '0')
			}
			layer[r] = row
		}
		image = append(image, layer)
	}
	return image
}

func part1(image [][][]int) int {
	res := 0
	fewestZeros := math.MaxInt
	var counts map[int]int
	for _, layer := range image {
		counts = make(map[int]int)
		for _, row := range layer {
			for _, n := range row {
				counts[n]++
			}
		}
		if counts[0] < fewestZeros {
			fewestZeros = counts[0]
			res = counts[1] * counts[2]
		}
	}
	return res
}

func part2(image [][][]int) string {
	h := len(image[0])
	w := len(image[0][0])
	colorMap := map[int]rune{0: ' ', 1: '█'}
	var sb strings.Builder
	var i, color int
	for r := range h {
		for c := range w {
			color = 2
			i = 0
			for color == 2 {
				color = image[i][r][c]
				i++
			}
			sb.WriteRune(colorMap[color])
		}
		sb.WriteRune('\n')
	}
	return sb.String()
}
